src/pybind/mgr/cephadm/nfs_ganesha.py

```diff
@@ -0,0 +1,114 @@
+import json
+
+from mgr_module import MgrModule
+from typing import Dict, Optional
+
+class NFSGanesha(object):
+    def __init__(self,
+                 mgr,
+                 daemon_id,
+                 pool,
+                 namespace=None):
+        # type: (MgrModule, str, str, Optional[str]) -> None
+        self.daemon_type = 'nfs'
+        self.daemon_id = daemon_id
+
+        self.mgr = mgr
+        self.log = mgr.log
+
+        # rados pool config
+        self.pool = pool
+        self.namespace = namespace
+
+        # TODO: add a const for this?
+        self.ceph_conf_path = '/etc/ceph/ceph.conf'
+
+    def get_daemon_name(self):
+        # type: () -> str
+        return '%s.%s' % (self.daemon_type, self.daemon_id)
+
+    def get_rados_user(self):
+        # type: () -> str
+        return 'admin' # TODO: 'nfs.' + self.daemon_id
+
+    def get_rados_config_obj(self):
+        # type: () -> str
+        return 'conf-' + self.get_daemon_name()
+
+    def get_rados_config_url(self):
+        # type: () -> str
+        url = 'rados://' + self.pool + '/'
+        if self.namespace:
+            url += self.namespace + '/'
+        url += self.get_rados_config_obj()
+        return url
+
+    def create_keyring(self):
+        # type: () -> str
+        self.log.debug('create nfs keyring for user: %s' % self.get_rados_user())
+        ret, keyring, err = self.mgr.mon_command({
+            'prefix': 'auth get',
+            'entity': 'client.' + self.get_rados_user(),
+        })
+        return keyring
+
+    def create_rados_config_obj(self):
+        # type: () -> None
+        obj = self.get_rados_config_obj()
+        self.log.debug('create rados config object: %s' % obj)
+        with self.mgr.rados.open_ioctx(self.pool) as ioctx:
+            if self.namespace:
+                ioctx.set_namespace(self.namespace)
+            ioctx.write_full(obj, ''.encode('utf-8'))
+
+    def get_ganesha_conf(self):
+        # type: () -> str
+        return '''# generated by cephadm
+NFS_CORE_PARAM {{
+        Enable_NLM = false;
+        Enable_RQUOTA = false;
+        Protocols = 4;
+}}
+CACHEINODE {{
+        Dir_Chunk = 0;
+        NParts = 1;
+        Cache_Size = 1;
+}}
+EXPORT_DEFAULTS {{
+        Attr_Expiration_Time = 0;
+}}
+NFSv4 {{
+        Delegations = false;
+        RecoveryBackend = 'rados_cluster';
+        Minor_Versions = 1, 2;
+}}
+RADOS_KV {{
+        ceph_conf = '{ceph_conf_path}';
+        userid = {userid};
+        nodeid = {nodeid};
+        pool = '{pool}';
+        namespace = '{namespace}';
+}}
+RADOS_URLS {{
+        ceph_conf = '{ceph_conf_path}';
+        userid = {userid};
+        watch_url = '{url}';
+}}
+%url    {url}
```

Comment:
@varshar16, @jtlayton, and I have been trying to distill the minimal config for Ganesha necessary for orchestrating Ganesha deployment. See #32297.

I think we can and should trim this down. The module adding NFS daemons should add to the Ganesha config (in RADOS) according to what the daemon will be doing.

I think here it could be:

````
RADOS_URLS {
 userid = {userid};
}

%url {url}
````

I deliberately removed `ceph_conf =...`. This is an unfortunately redundant variable in three sections: `RADOS_URLS` (ganesha conf from RADOS), `RADOS_KV` (FSAL_CEPH recovery database for CephFS), and `FSAL_CEPH` (CephFS proxy). I think the best way to do this is an environment variable for the nfs-ganesha process, `CEPH_CONF`. Is that feasible to add cephadm? cc @liewegas 

I'm not expert in this code. How does a module (like `pybind/mgr/volumes`) do run-time configuration of the Ganesha daemon via the RADOS object? What about updating the caps on the `{userid}`?

Reply:
> @varshar16, @jtlayton, and I have been trying to distill the minimal config for Ganesha necessary for orchestrating Ganesha deployment. See #32297.
> 
nice!  :+1: 

> I think we can and should trim this down. The module adding NFS daemons should add to the Ganesha config (in RADOS) according to what the daemon will be doing.
> 
> I think here it could be:
> 
> ```
> RADOS_URLS {
>  userid = {userid};
> }
> 
> %url {url}
> ```
> 

Agree, I'll look at trimming this down and/or for what we can leverage from #32297

> I deliberately removed `ceph_conf =...`. This is an unfortunately redundant variable in three sections: `RADOS_URLS` (ganesha conf from RADOS), `RADOS_KV` (FSAL_CEPH recovery database for CephFS), and `FSAL_CEPH` (CephFS proxy). I think the best way to do this is an environment variable for the nfs-ganesha process, `CEPH_CONF`. Is that feasible to add cephadm? cc @liewegas
> 

a `CEPH_CONF` environment var shouldn't be a problem .. I can add it to PR #33064



> I'm not expert in this code. How does a module (like `pybind/mgr/volumes`) do run-time configuration of the Ganesha daemon via the RADOS object?
>

Can `pybind/mgr/volumes` use the orchestrator `ServiceSpec` ..?
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/orchestrator/_interface.py#L1474

this is how the dashboard integrates:
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/services/ganesha.py#L124

> What about updating the caps on the `{userid}`?
>
yes, if possible I'd rather get the caps from the shared logic in #32297

Reply:
@batrick 

PR #33064 has been modified to include:
* a `CEPH_CONF` environment var: https://github.com/ceph/ceph/pull/33064/commits/635adb5de19e37b1d7c4ce709019eb95ddf5e30e
* a trimmed ganesha.conf: https://github.com/ceph/ceph/commit/ada3aa81afd3fd300f127872a067c64026991586

I'll fix-up this PR next ...